pkg/apis/messaging/v1alpha1/choice_types.go

```diff
@@ -0,0 +1,165 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"knative.dev/pkg/apis"
+	duckv1alpha1 "knative.dev/pkg/apis/duck/v1alpha1"
+	duckv1beta1 "knative.dev/pkg/apis/duck/v1beta1"
+	"knative.dev/pkg/webhook"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// Choice defines conditional branches that will be wired in
+// series through Channels and Subscriptions.
+type Choice struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the Choice.
+	Spec ChoiceSpec `json:"spec,omitempty"`
+
+	// Status represents the current state of the Choice. This data may be out of
+	// date.
+	// +optional
+	Status ChoiceStatus `json:"status,omitempty"`
+}
+
+// Check that Choice can be validated, can be defaulted, and has immutable fields.
+var _ apis.Validatable = (*Choice)(nil)
+var _ apis.Defaultable = (*Choice)(nil)
+
+// TODO: make appropriate fields immutable.
+//var _ apis.Immutable = (*Choice)(nil)
+var _ runtime.Object = (*Choice)(nil)
+var _ webhook.GenericCRD = (*Choice)(nil)
+
+type ChoiceSpec struct {
+	// Cases is the list of Filter/Subscribers pairs. Filters are evaluated in the order
+	// provided, until one pass (returns true)
+	Cases []ChoiceCase `json:"cases"`
+
+	// ChannelTemplate specifies which Channel CRD to use
+	ChannelTemplate *eventingduckv1alpha1.ChannelTemplateSpec `json:"channelTemplate"`
+
+	// Reply is a Reference to where the result of a case Subscriber gets sent to
+	// when the case does not have a Reply
+	//
+	// You can specify only the following fields of the ObjectReference:
+	//   - Kind
+	//   - APIVersion
+	//   - Name
+	//
+	//  The resource pointed by this ObjectReference must meet the Addressable contract
+	//  with a reference to the Addressable duck type. If the resource does not meet this contract,
+	//  it will be reflected in the Subscription's status.
+	// +optional
+	Reply *corev1.ObjectReference `json:"reply,omitempty"`
+}
+
+type ChoiceCase struct {
+	// Filter is the expression guarding the branch/case
+	Filter *eventingv1alpha1.SubscriberSpec `json:"filter,omitempty"`
+
+	// Subscriber receiving the event when the filter passes
+	Subscriber eventingv1alpha1.SubscriberSpec `json:"subscriber"`
+
+	// Reply is a Reference to where the result of Subscriber of this case gets sent to.
+	// If not specified, sent the result to the Choice Reply
+	//
+	// You can specify only the following fields of the ObjectReference:
+	//   - Kind
+	//   - APIVersion
+	//   - Name
+	//
+	//  The resource pointed by this ObjectReference must meet the Addressable contract
+	//  with a reference to the Addressable duck type. If the resource does not meet this contract,
+	//  it will be reflected in the Subscription's status.
+	// +optional
+	Reply *corev1.ObjectReference `json:"reply,omitempty"`
```

Comment:
Does it make sense for choice to support multi-leg Subscriber/Reply?

Reply:
I was going back and forth on this one and I think one good use case is error-handling, which can be done by having one leg sending the reply to an error sink when the event is an error (TBD).